README.markdown
# Zoomable Type: Mac OS X’s “Show in Large Type” for jQuery

## What

Zoomable Type is a jQuery plugin for reading small text on a display from far away. Text is upscaled to the largest size based on its width and the browser's viewport. [Try live examples](http://zoomabletype.tatey.com/).

[![Preview](https://cloud.github.com/downloads/tatey/jquery-zoomabletype/zoomed_preview.png)](http://zoomabletype.tatey.com/)

## Licence

Distributed under the MIT licence. See LICENCE.


CHANGELOG.markdown
# Changelog 

## v1.0.4 / 2011-04-17
  * Massive refactoring
  * Disable on all mobile devices, not just iPhone

## v1.0.3 / 2010-04-26
  * Use jQuery selector instead of element ID or class.
  * zoomableType() is chainable.
  * zoomableType() accepts an optional argument to set CSS
    class on zoomed element. Defaults to zoomed.

## v1.0.2 / 2010-03-16
  * Hide zoomed type on keydown.
  * Don't bind to elements on iPhone.

## v1.0.1 / 2010-02-26
  * Fixed compatibility with Safari 4.

## v1.0.0 / 2010-02-25
  * First release.


jquery.zoomabletype.js
/*
 * Zoomable Type: Mac OS X's "Show in Large Type" for jQuery
 *
 * @version 1.0.4
 * @date    2011-04-17
 * @require jQuery >= v1.5.2 
 * @www     http://zoomabletype.tatey.com/
 *
 * Zoomable Type is a jQuery plugin for reading small text on a
 * display from far away.
 *
 * Distributed under the MIT licence:
 * See LICENCE or http://www.opensource.org/licenses/mit-license.php
 * 
 * Copyright 2010 Tate Johnson <tate@tatey.com>
 */

(function($) {
  var ZoomableType = {
    _doc: $(document),
    _win: $(window),
    _isZoomed: false,
    
    Element: function(element, options) {
      this.options  = $.extend({className: 'zoomed', textShadowColor: '#000'}, options);
      this.copy     = null;
      this.original = $(element).bind('click.zt', this, function(e) {
        if (ZoomableType._isZoomed) return;
        e.data.zoom();
        e.stopPropagation();
      });
      
      return this;
    }
  }

  $.extend(ZoomableType.Element.prototype, {
    zoom: function() {
      this.copy = $('<span>', {className: this.options.className})
        .text(this.original.text())
        .appendTo('body')
        .bind('click.zt', function(e) { e.stopPropagation(); });
      this.resize();
      
      ZoomableType._isZoomed = true;
      ZoomableType._doc.bind('click.zt keyup.zt', this, function(e) { 
        if (!e.keyCode || e.keyCode == 27) e.data.shrink();
      });
      ZoomableType._win.bind('resize.zt', this, function(e) { 
        e.data.resize(); 
      });
    },
    
    shrink: function() {      
      this.copy.remove();
      this.copy = null;
      
      ZoomableType._isZoomed = false;
      ZoomableType._doc.unbind('.zt');
      ZoomableType._win.unbind('.zt');
    },
    
    resize: function() {
      var width  = ZoomableType._win.width() * 90 / this.original.width(),
          height = ZoomableType._win.height() * 90 / this.original.height();
      
      this.copy.css('font-size', (width < height ? width : height) + '%');
      this._style()._center();
    },
        
    _style: function() {
      var radius = this.copy.outerWidth() * 0.02;
      
      this.copy.css({
        padding: radius,
        webkitBorderRadius: radius,
        mozBorderRadius: radius,
        borderRadius: radius,
        textShadow: this.options.textShadowColor + ' 2px 2px ' + radius + 'px'
      });
      
      return this;
    },
    
    _center: function() {
      this.copy.css({
        position: 'fixed',
        left: (ZoomableType._win.width() - this.copy.outerWidth()) / 2,
        top: (ZoomableType._win.height() - this.copy.outerHeight()) / 2
      });
      
      return this;
    }
  });
    
  $.fn.zoomabletype = function(options) {
    if (window.screen.width > 480) {
      this.each(function() { new ZoomableType.Element(this, options); });
    }
    return this;
  }
})(jQuery);


zoomable_type.jquery.js
/*
 * Zoomable Type: Mac OS X's "Show in Large Type" for jQuery
 *
 * @version 1.0.3
 * @date    2010-04-26
 * @require jQuery >= v1.4.2 
 * @www     http://zoomabletype.tatey.com/
 *
 * Zoomable Type is a jQuery plugin for reading small text on a
 * display from far away.
 *
 * Distributed under the MIT licence:
 * See LICENCE or http://www.opensource.org/licenses/mit-license.php
 * 
 * Copyright (c) 2010 Tate Johnson <tate@tatey.com>
 */

(function($) {  
  $.fn.zoomableType = function(klass) {
    if (navigator.userAgent.match(/iPhone/)) return this; // Mobile Safari on iPhone already binds event listeners to "telephone numbers"
    $(this).live('click', function() {
      if ($.zoomableType) return false; // There can only be one. Let document event listener clean up.
      $.zoomableType = $('<span>', {'id': 'zt' + new Date().getTime(), 'class': (klass || 'zoomed')})
        .text($(this).text())
        .appendTo('body')
        .attr('data-unzoomed-width', $(this).width())
        .attr('data-unzoomed-height', $(this).height());
      setSizeAndPosition($.zoomableType);
      $(document).bind('click.zt keydown.zt', unzoom);
      $(window).bind('resize.zt', setSizeAndPosition);
      
      /* 
       * Unbind event listeners and remove zoomed element from DOM.
       */
      function unzoom() {
        $(document).unbind('click.zt keyup.zt');
        $(window).unbind('resize.zt');
        $.zoomableType.remove();
        $.zoomableType = null;
      }
      
      /*
       * Zoom text to 90% of browser window's width or height (Whichever is greater). 
       * Position zoomed text in centre of the window.
       */
      function setSizeAndPosition() {
        var win         = $(window);
        var widthScale  = (win.width() - win.width() * 0.10) / $.zoomableType.attr('data-unzoomed-width');
        var heightScale = (win.height() - win.height() * 0.10) / $.zoomableType.attr('data-unzoomed-height');
        $.zoomableType.css(
          'font-size',
          (widthScale < heightScale ? widthScale : heightScale) * 100 + '%'
        );
        var radius = $.zoomableType.outerWidth() * 0.02;
        $.zoomableType.css({
          'padding': radius,
          '-moz-border-radius': radius,
          '-webkit-border-radius': radius,
          'border-radius': radius,
          'text-shadow': '#000 2px 2px ' + parseFloat($.zoomableType.css('fontSize')) * 0.01 + 'px'
        });
        // Position after styling has been applied.
        $.zoomableType.css({
          'left': (win.width() - $.zoomableType.outerWidth()) / 2,
          'top': win.height() / 2 - $.zoomableType.outerHeight() / 2
        });
      }
    });
    return this;
  }
})(jQuery);


jquery.zoomable_type.js
/*
* Zoomable Type: Mac OS X's "Show in Large Type" for jQuery
* @version 1.0.2
* @date 2010-03-16
* @require jQuery >= v1.4.2 
*
* Zoomable Type is a jQuery plugin for reading small text on a
* display from far away.
*
* Distributed under the MIT licence:
* http://www.opensource.org/licenses/mit-license.php
*
* Copyright (c) 2010 Tate Johnson <tate@tatey.com>
*/

(function($) {  
  $.zoomableType = function(selector, settings) {
    if (!navigator.userAgent.match(/iPhone/)) {
      $(selector || '.zoomable').each(function() { $(this).click(onZoom); });
    }
  };
  var $zt = $.zoomableType;
  
  function onZoom() {
    if ($zt.zoomedElement) {
      onHide();
      return false; 
    }
    var element = $('<span>', {'id': 'zoomed' + (+ new Date()), 'class': 'zoomed'});
    element.text($(this).text());
    element.appendTo('body');
    element.data.unzoomedWidth  = element.width();
    element.data.unzoomedHeight = element.height();
    setSizeAndPositionFor(element);
    $zt.zoomedElement = element;
    $(document).bind('click keydown', onHide);
    $(window).resize(onWindowResize);
    return false;
  }
  
  function onHide() {
    $(document).unbind('click keydown', onHide);
    $(window).unbind('resize', onWindowResize);
    $zt.zoomedElement.remove();
    $zt.zoomedElement = null;
    return false;
  }
  
  function onWindowResize() {
    setSizeAndPositionFor($zt.zoomedElement);
  }
  
  function setSizeAndPositionFor(element) {
    var win         = $(window);
    var widthScale  = (win.width() - win.width() * 0.10) / element.data.unzoomedWidth;
    var heightScale = (win.height() - win.height() * 0.10) / element.data.unzoomedHeight;
    element.css(
      'font-size',
      (widthScale < heightScale ? widthScale : heightScale) * 100 + '%'
    );
    var radius = element.outerWidth() * 0.02;
    element.css({
      'padding': radius,
      '-moz-border-radius': radius,
      '-webkit-border-radius': radius,
      'border-radius': radius,
      'text-shadow': '#000 2px 2px ' + parseFloat(element.css('fontSize')) * 0.01 + 'px'
    });
    element.css({
      'left': (win.width() - element.outerWidth()) / 2,
      'top': win.height() / 2 - element.outerHeight() / 2
    });
  }
})(jQuery);
